frontend/catalyst/debug/compiler_functions.py

```diff
@@ -202,3 +208,92 @@ def replace_ir(fn, stage, new_ir):
     fn.overwrite_ir = new_ir
     fn.compiler.options.checkpoint_stage = stage
     fn.fn_cache.clear()
+
+
+@debug_logger
+def compile_executable(fn, *args):
+    """Generate and compile a C program that calls a jitted function with the provided arguments.
+
+    Args:
+        fn (QJIT): a qjit-decorated function
+        *args: argument values to use in the C program when invoking ``fn``
+
+    Returns:
+        (str): the paths that should be included in LD_LIBRARY_PATH.
+        (str): the path of output binary.
+    """
+    # if fn is not compiled, compile it first.
+    if not fn.compiled_function:
+        fn(*args)
+
+    # get python version
+    python_lib_dir_path = sysconfig.get_config_var("LIBDIR")
+    version_info = sys.version_info
+
+    # If libpython3.so exists, link to that instead of libpython3.x.so
+    if os.path.isfile(python_lib_dir_path + f"/libpython{version_info.major}.so"):
+        version_str = f"{version_info.major}"
+    else:
+        version_str = f"{version_info.major}.{version_info.minor}"
+
+    lib_path_flags = [
+        f"-Wl,-rpath,{python_lib_dir_path}",
+        f"-L{python_lib_dir_path}",
+        "-lpython" + version_str,
+    ]
+
+    # Linker in macOS might use @rpath/Python3.framework/Versions/3.x/Python3.
+    if platform.system() == "Darwin":  # pragma: nocover
+        python_lib_dir_rpath = python_lib_dir_path.split("Python3.framework")[0]
+        lib_path_flags.insert(1, f"-Wl,-rpath,{python_lib_dir_rpath}")
+
+    f_name = str(fn.__name__)
+    workspace = str(fn.workspace) if fn.compile_options.keep_intermediate else os.getcwd()
+    main_c_file = workspace + "/main.c"
+    output_file = workspace + "/" + f_name + ".out"
+    shared_object_file = workspace + "/" + f_name + ".so"
+
+    # copy shared object to current directory
+    if not fn.compile_options.keep_intermediate:
+        original_shared_object_file = str(fn.workspace) + "/" + f_name + ".so"
+        shutil.copy(original_shared_object_file, shared_object_file)
+
+    options = fn.compiler.options
+    with open(main_c_file, "w", encoding="utf-8") as file:
+        file.write(get_cmain(fn, *args))
+
+    # configure flags
+    default_flags = LinkerDriver.get_default_flags(options)
+    no_shared_flags = [fs for fs in default_flags if fs != "-shared"]
+    link_so_flags = (
+        no_shared_flags
+        + [
+            "-Wl,-rpath," + workspace,
+            shared_object_file,
+        ]
+        + lib_path_flags
+    )
```

Comment:
I'm still a bit hesitant about reusing the entire flags used for the shared object, normally those should be transitively used when loading that library. Could you report what exactly went wrong when you use only the relevant flags (that we already generate here anyways)?

Reply:
If we do not include the flags, we will get many undefined references errors, like theses.
```shell
/lib/python3.12/site-packages/scipy/../scipy.libs//libopenblasp-r0-23e5df77.3.21.dev.so, not found (try using -rpath or -rpath-link)
/lib/python3.12/site-packages/scipy/../scipy.libs//libopenblasp-r0-23e5df77.3.21.dev.so: undefined reference to `_gfortran_etime@GFORTRAN_8'
...
``` 
I will try looking into how to get these libraries loaded without duplicating the flags.

Reply:
I think that means the rpath is not set correctly in the `f_name.so` (where the openblas dependency is declared)? 🤔

Reply:
Including `-lgfortran` solves the issue. Right now we can compile executable without these flags.

Reply:
It looks like we have to add gfortran and quadmath into default flag so that compilation with mcin.c can find them.

Reply:
That sounds like the rpath fix is not working?

Reply:
Hey folks, looking here: As long as the `$ORIGN` (or `loader_path` on MacOS), and subsequent paths are set properly in the rpath of the resulting library/binary, there should be no need to add gfortran and quadmath, as those are local to the scipy wheel libs dir.

Reply:
Adding ``$ORIGIN`` to executable compilation works. Will test MacOS.